modules/cert-manager-acme-http01.adoc

```diff
@@ -0,0 +1,108 @@
+// Module included in the following assemblies:
+//
+// * security/cert_manager_operator/cert-manager-operator-issuer-acme.adoc
+
+:_content-type: PROCEDURE
+[id="cert-manager-acme-http01_{context}"]
+// = Using the HTTP01 Ingress solver
+
+= Issuing certificates by using the HTTP01 challenge
+// TODO: Name
+
+TODO: intro
+
+.Prerequisites
+
+* You have access to the cluster as a user with the `cluster-admin` role.
+* TODO
+// * User workload to expose
+// TODO: ^ from gdoc, what does this mean?
+
+.Procedure
+
+. Create an ACME cluster issuer.
+
+.. Create a YAML file that defines the `ClusterIssuer` object:
++
+.Example `acme-cluster-issuer.yaml` file
+[source,yaml]
+----
+apiVersion: cert-manager.io/v1
+kind: ClusterIssuer
+metadata:
+  name: letsencrypt-prod                                           <1>
+spec:
+  acme:
+    preferredChain: ""
+    privateKeySecretRef:
+      name: example-issuer-account-key-prod                        <2>
+    server: https://acme-staging-v02.api.letsencrypt.org/directory <3>
+    solvers:
+      - http01:
+          ingress:
+            class: openshift-default                               <4>
+----
+<1> TODO: Need description or know if they need to change.
+<2> TODO: Need description or know if they need to change.
+<3> TODO: Need description or know if they need to change.
+<4> TODO - Specify `openshift-default` to use the Ingress-to-Route controller (for what purpose).
+
+
+.. Create the `ClusterIssuer` object by running the following command:
++
+[source,terminal]
+----
+$ oc create -f acme-cluster-issuer.yaml
+----
+
+. Create an ingress to expose the service of the user workload.
+
+.. Create a YAML file that defines the `Ingress` object:
++
+.Example `ingress.yaml` file
+[source,yaml]
+----
+apiVersion: networking.k8s.io/v1
+kind: Ingress
+metadata:
+  name: sample-ingress                                           <1>
+  namespace: sandbox                                             <2>
+  annotations:
+    cert-manager.io/cluster-issuer: letsencrypt-prod             <3>
+    acme.cert-manager.io/http01-ingress-class: openshift-default <4>
+spec:
+  ingressClassName: openshift-default                            <5>
```

Comment:
@thejasn @swghosh Can you give more details about this line and callout 4? Is callout 5 referencing the "openshift-default" class we put in the cluster issuer above? 

And then what is the annotation in callout 4 needed for?

Reply:
<5> : Is part of kubernetes networking API, https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.22/#ingressspec-v1-networking-k8s-io. As to why the class name is `openshift-default` is described [here](https://www.google.com/url?q=https://docs.google.com/document/d/18A90OEC6nWZocqdzzRs1CUpvL39InaI5q0IAeuI21tc/edit&sa=D&source=docs&ust=1678188022619622&usg=AOvVaw1rNGlx2As9_0hl5d9K61AN)

<4>: Is part of upstream cert-manager, https://cert-manager.io/docs/usage/ingress/#supported-annotations
> acme.cert-manager.io/http01-ingress-class: this annotation allows you to configure the ingress class that will be used to solve challenges for this ingress. Customizing this is useful when you are trying to secure internal services, and need to solve challenges using a different ingress class to that of the ingress. If not specified and the acme-http01-edit-in-place annotation is not set, this defaults to the ingress class defined in the Issuer resource.

The value basically needs to correlate to the ingress class that is being used. `openshift-default` is automatically created as part of the `default` ingresscontroller. Hence the `default` suffix in the ingressclass name.

Reply:
@bergerhoffer @thejasn given the previously created cluster issuer already specified ingress class:
```
      - http01:
          ingress:
            class: openshift-default
```
AND the ingress resource already specified "cert-manager.io/cluster-issuer: letsencrypt-staging" to use the previously created cluster issuer, then this is enough to be a normal usage, it will function well, i.e. these two fields 'acme.cert-manager.io/http01-ingress-class: openshift-default' and 'ingressClassName: openshift-default' can be not needed, (but no hurt if these two exist too I think) @thejasn I suggest remove these two, WDYT?

you can see my https://docs.google.com/document/d/1ZUzGBIZVOwqQrEB-M0zE9aAjK8_ue7Gfsxv_jm4XX7Q the "http01 solver" section's first part about integration with ingress, need not specify 'acme.cert-manager.io/http01-ingress-class: openshift-default' and 'ingressClassName: openshift-default'

Reply:
Without `'ingressClassName: openshift-default'`, a Route won't be created for the created Ingress. We will require this field. Can you verify if a route was created for the ingress in your tests? ([Additional info](https://docs.google.com/document/d/18A90OEC6nWZocqdzzRs1CUpvL39InaI5q0IAeuI21tc/edit))

Reply:
@bergerhoffer @thejasn , yes, routes are created, note there are 2 routes created, see my https://docs.google.com/document/d/1ZUzGBIZVOwqQrEB-M0zE9aAjK8_ue7Gfsxv_jm4XX7Q/ the "http01 solver" section's first part's detailed steps and results.

Reply:
Hmm looks like the field is not required by the `ingress-to-route` controller ([ref](https://github.com/openshift/route-controller-manager/blob/d7a8e22db412b6fabb7028ca0da8de8f3d9ac3c3/pkg/route/ingress/ingress.go#L572-L595)) since it defaults to managing any Ingress that doesn't have `ingressClassName` specified. We may be fine not using this field if the user wants to use the default ingress controller. Since the route will be admitted by the default router in this case. 

If the user wants to use a different (secondary) ingresscontroller (meaning the route to be only admitted by the second router), they would need to specific `ingressClassName`  to be `openshift-secondary`. Agreed that this also requires additional changes to the ingresscontroller (router selectors, etc) but these are documented under ingresscontroller docs. For us the scope ends at `ingressClassName` I think.

Reply:
Yes. I also didn't have time to investigate ingress class, ingress controller knowledge for the secondary ingress class name's test. I agree our doc needs not touch secondary ingress class name for now. If needed for future, we can have more time to test/document it then.

Reply:
Okay so for now I'm leaving the openshift-default values in all 3 places, and am updating the descriptions to very generic "Specify the ingress class." Please let me know if any of the descriptions should be enhanced